File: library/src/main/java/io/github/wax911/library/model/request/QueryContainer.java
package io.github.wax911.library.model.request;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Map;
import java.util.WeakHashMap;

/**
 * Actual query and variable container
 * used in retrofit requestBodyConverter
 */
public class QueryContainer implements Parcelable {

    private String query;
    private Map<String, Object> variables;

    QueryContainer() {
        variables = new WeakHashMap<>();
    }

    @SuppressWarnings("unchecked")
    private QueryContainer(Parcel in) {
        query = in.readString();
        variables = in.readHashMap(WeakHashMap.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(query);
        dest.writeMap(variables);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<QueryContainer> CREATOR = new Creator<QueryContainer>() {
        @Override
        public QueryContainer createFromParcel(Parcel in) {
            return new QueryContainer(in);
        }

        @Override
        public QueryContainer[] newArray(int size) {
            return new QueryContainer[size];
        }
    };

    protected void setQuery(String query) {
        this.query = query;
    }

    void putVariable(String key, Object value) {
        variables.put(key, value);
    }

    public Map<String, Object> getVariables() {
        return variables;
    }

    boolean containsVariableKey(String key) {
        return variables != null && variables.containsKey(key);
    }
}

File: library/src/main/kotlin/io/github/wax911/library/model/request/QueryContainer.kt
/**
 * Copyright 2021 AniTrend
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.wax911.library.model.request

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.RawValue
import java.util.WeakHashMap

/**
 * GraphQL request that follows the common GraphQL HTTP request format
 *
 * @see [GraphQL Over HTTP](https://graphql.org/learn/serving-over-http/#post-request)
*/
@Parcelize
data class QueryContainer internal constructor(
    var operationName: String? = null,
    var query: String? = null,
    val variables: @RawValue MutableMap<String, Any?> = WeakHashMap(),
    val extensions: @RawValue MutableMap<String, Any?> = WeakHashMap(),
) : Parcelable {
    internal fun putVariables(values: Map<String, Any?>) {
        variables.putAll(values)
    }

    internal fun putVariable(
        key: String,
        value: Any?,
    ) {
        variables[key] = value
    }

    internal fun putExtension(
        key: String,
        value: Any?,
    ) {
        extensions[key] = value
    }
}
